CMakeLists.txt

```diff
@@ -217,47 +217,80 @@ add_test(
+    # Expose `${TARGET_NAME}` and its dependency to the parent scope for further use
+    set(${TARGET_NAME}_DEP ${TARGET_NAME} ${OUTPUT_FILEPATH} PARENT_SCOPE)
 endfunction()
 
-list(TRANSFORM JS_DIRS REPLACE "^[^:]+:([^:]+)$" \\1 OUTPUT_VARIABLE JS_DEPS)
-file(GLOB common LIST_DIRECTORIES false ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common/*)
-file(GLOB common-fonts LIST_DIRECTORIES false ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common/fonts/*)
+# Static (not built) resource dependencies
+file(GLOB common-misc-resources LIST_DIRECTORIES false ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common/resources/*)
+file(GLOB common-fonts LIST_DIRECTORIES false ${CMAKE_CURRENT_SOURCE_DIR}/services/user/CommonSys/common/resources/fonts/*)
 
-# Builds the Common Sys Lib
-add_js_target(${COMMON_DIR} common-lib.js)
+# Builds the Common Sys files that the UIs (just below) depend on
+add_js_target(${COMMON_LIB_DIR} common-lib.js)
+add_js_target(${COMMON_LOADER_DIR} index.html ${CommonSysCommonLib_js_DEP})
 
-foreach(ENTRY ${JS_DIRS})
-    add_js_target(${ENTRY} index.html ${common} ${common-fonts} CommonSysLib_js)
+# Build each UI
+# Each of these UIs receives the build output of CommonSys, so CommonSys files are listed as dependencies
+foreach(ROW ${JS_DIRS})
+    add_js_target(${ROW} index.html ${common-misc-resources} ${common-fonts} ${CommonSysCommonLib_js_DEP} ${CommonSysLoaderLib_js_DEP})
 endforeach()
 
+# A function that can be used to add a new target that builds a wasm component from rust
+set(COMPONENT_BIN_DIR ${CMAKE_CURRENT_BINARY_DIR}/components)
+function(add_rs_component TARGET_TUPLE OUTPUT_FILE)
+    string(REGEX REPLACE "^([^:]+):([^:]+)$" \\1 PATH ${TARGET_TUPLE})
+    string(REGEX REPLACE "^([^:]+):([^:]+)$" \\2 TARGET_NAME ${TARGET_TUPLE})
+    set(OUTPUT_FILEPATH ${COMPONENT_BIN_DIR}/${OUTPUT_FILE})
+    file(GLOB_RECURSE SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/src/*)
+
+    # Add a command that rebuilds `${OUTPUT_FILEPATH}` whenever any sources or dependencies are updated
+    add_custom_command(
+        OUTPUT ${OUTPUT_FILEPATH}
+        DEPENDS ${SOURCES} ${ARGN}
+        COMMAND cargo component build 
+            --release 
+            --target wasm32-unknown-unknown
+            --manifest-path ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/Cargo.toml 
+            --target-dir    ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/target
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${COMPONENT_BIN_DIR}
+        COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/${PATH}/target/wasm32-unknown-unknown/release/${OUTPUT_FILE} ${OUTPUT_FILEPATH}
+    )
+    # Create a custom target to ensure `${OUTPUT_FILEPATH}` is built before `${TARGET_NAME}`
+    add_custom_target(${TARGET_NAME} DEPENDS ${OUTPUT_FILEPATH})
+    # Expose `${TARGET_NAME}` and its dependency to the parent scope for further use
+    set(${TARGET_NAME}_DEP ${TARGET_NAME} PARENT_SCOPE)
```

Comment:
When used in other custom commands, this adds only a target-level dependency. A file-level dependency is also required to ensure that out-of-dateness propagates correctly.

Reply:
fixed